fix spam filter (#33)
opst/knitfab-docs: 04.examples/news-classification/news-classification.ja.md

# Example: Knitfab での LLM のファインチューニングによるニュース分類

> 本書では、Knitfab を使用した一般大規模言語モデル (LLM) をファインチューニングすることにより、モデルがニュースといった特定の文書分類性能を改善できることを示します。この事例を通して、Knitfab がファインチューニング実験の入力と出力、および履歴の管理を自動化しながら、LLM-as-a-judge を用いて効率的に人間のような評価ができるようになります。

Reviewer: ```suggestion
本書では、Knitfab を使用して、大規模言語モデル (LLM) をファインチューニングすることにより、ニュースといった特定の文書についての分類性能を改善できることを示します。この事例を通して、Knitfab がファインチューニング実験の入力と出力、および履歴の管理を自動化しながら、LLM-as-a-judge を用いて効率的に人間のような評価ができることを示します。
```

主語と述語の関係が食い違っているのに気が付きました。

## 概要

> LLM のファインチューニングは、特定のタスクに対して処理性能と判断性能を高めるために複数なハイパーパラメータを調整する工程です。しかしながら、複雑なハイパーパラメータにより LLM のファインチューニングの管理はすぐに難しいものになります。そこで、 Knitfab により学習工程にくわえて学習の入出力、成果物まで自動管理することで、さまざまな構成設定での実験が容易になります。さらに、革新的な評価方法である LLM-as-a-judge を用いることにより人間のような評価品質を維持しながらコストの削減も可能になります。本事例では、このニュース分類タスクのファインチューニングから LLM-as-a-judge 評価までの一連のタスクを Knitfab 上で実装することを体験し、リネージ自動管理プラットフォームとして Knitfab の強みと容易さをお伝えします。

Reviewer: ```suggestion
LLM のファインチューニングでは、特定のタスクに対する処理性能と判断性能を高めるために、モデルの再訓練を行う工程です。しかしながら、ハイパーパラメータなど、訓練条件は様々に設定できるので、それら条件を用いて試行錯誤をする中でファインチューニングの管理はすぐに難しいものになります。そこで、 Knitfab を用いることで、学習工程にくわえて学習の入出力、成果物まで自動管理できるようになり、さまざまな条件での実験管理が容易になります。さらに本事例では、革新的な評価方法である LLM-as-a-judge を Knitfab と連携させる手法を示します。LLM-as-a-judge により、人間のような評価品質を維持しながらコストの削減も可能になります。本事例では、このニュース分類タスクのファインチューニングから LLM-as-a-judge 評価までの一連のタスクを Knitfab 上で実装することを体験し、リネージ自動管理プラットフォームとして Knitfab の強みと容易さをお伝えします。
```

前半については、論理的整合性を見直しました。

ところで、後半の「さらに」以降ですが、この脈絡が前半と噛み合っていないように感じましたので、少々手を加えました。

Author: リネージ自動管理プラットフォーム -> 機械学習パイプラインの自動制御プラットフォームの修正抜け漏れがありましたので、併せて修正いたします。

本章では、ニュース分類ファインチューニング済みのモデルを評価するための、LLM を使用した評価パイプラインについて説明します。`scripts/evaluate/evaluate.py` の `TestGPT2Model` クラスは、定義したテストケースとニュースカテゴリ分類精度指標に従ってモデルを評価し、分類性能レポートを生成します。これにより、モデル間に一貫したベンチマークに従って比較できるため、モデルの精度向上に寄与します。

### 2.1. 引数指定

> `parse_arguments` 関数は、評価プロセスの必要な引数を処理します。さまざまなパラメータを任意に指定できるため、モデルのファインチューニングに柔軟に対応で来ます。

Reviewer: ```suggestion
`parse_arguments` 関数は、評価プロセスの必要な引数を処理します。さまざまなパラメータを任意に指定できるため、モデルのファインチューニングに柔軟に対応できます。
```

誤変換に気が付きました。（"be able to" の意味で「出来る」と書くことはありますが、「で来る」と書くことはありません。また、いずれにせよ、ひらがなに開いたほうがよいでしょう）

## ステップ 4:（任意）Docker イメージの動作確認
> [!Note]
>

> > イメージの動作確認は時間がかかますので、必要に応じて実施してください。途中で中断しても構いません。

Reviewer: ```suggestion
> イメージの動作確認としてタスクをすべて実施するとは時間がかかりますので、適宜中断してください。
```

本章では、Knitfab 上 LLM のニュース分類ファインチューニングを行います。

### 6.1. YAML ひな型の生成

YAML ひな型を生成するには、2つの選択肢があります。

- 選択肢①: 空のひな型を作成する
```bash
knit plan template --scratch > ./plans/news-classification-train.v1.0.yaml
```

> これにより、`./plans` ディレクトリに`news-classification-train.v1.0.yaml` という名前で新しい YAML ひな型が作成されます。このひな型の設定項目の値は全て「空（空欄）」になっていますので、ユーザが必要な項目を追記します。

Reviewer: ```suggestion
これにより、`./plans` ディレクトリに`news-classification-train.v1.0.yaml` という名前で新しい YAML ひな型が作成されます。このひな型の設定項目の値は全て空欄や仮の値になっていますので、ユーザが必要な項目を追記します。
```

knit run show --log $evaluate_run_id
```

### 7.8.（任意）指標レポートのダウンロード
```bash
knit data pull -x $evaluate_metrics_knit_id ./out
```
このコマンドは、評価指標レポートを Knitfab からダウンロードし、`./out` ディレクトリに保存します。

## ステップ8: パイプライン構造とリネージグラフの確認
このステップでは、構築したパイプライン構造を可視化し、データ依存関係を確認します。

### 8.1. Plan グラフの生成（パイプライン構造）
Plan グラフは、パイプラインの全体像を把握し、初期学習、増分学習、検証などの異なる Plan 間の依存関係を示します。

次のコマンドを実行して、Plan グラフの PNG 画像を生成してください。`${plan_id}` は、該当する Plan Id(`$train_plan_id`、`$evaluate_plan_id`) に置き換えてください。

```bash
knit plan graph -n all ${plan_id} | dot -Tpng > plan-graph.png
```

生成された `plan-graph.png` を確認し、以下の点を検証してください。

> - **下流依存関係:**
>   `train_plan` が、 `evaluate_plan` を下流タスクとして示していることを確認します。これは学習タスクの後に検証が実行されることを意味します。
> - **上流依存関係:**
>   `evaluate_plan` が `train_plan`を上流タスクとして示していることを確認します。これは、検証タスクが学習タスクから生成された `/out/model` を正しく受け取っていることを意味します。

Reviewer: 同じ edge について 2 つの方向から説明していますね。ひとつにまとめてシンプルにしましょう。

```suggestion
- **訓練タスクと評価タスクの関係:**
  - `train_plan` が、 `evaluate_plan` を下流タスクとして示していることを確認します。
  - これは次のことを意味します。
    - 学習タスクの後に検証が実行されること。
    - 検証タスクが学習タスクから生成された `/out/model` を受け取ること。
```

> 

Reviewer: ここに Plan Graph の画像を貼ってください。

### 7.2. リネージグラフの生成（データフロー）

リネージグラフは、パイプライン内のデータと成果物の流れを遡ります。

以下のコマンドを実行して、リネージグラフの PNG 画像（`lineage-graph.png`）を生成してください。`${knit_id}` は、該当する Knit Id (`$train_model_knit_id`、`$evaluate_metrics_knit_id`) に置き換えてください。

```bash
knit data lineage -n all ${knit_id} | dot -Tpng > lineage-graph.png
```
生成された `lineage-graph.png` を確認し、以下の点を検証してください。

- **学習の出力:**
  `$train_run` が `/out/model`（学習済みモデル）と `(log)`（実行ログ）を出力として生成していることを確認します。
- **検証の出力:**

>   `$evaluate_run` が、学習タスクからの出力モデルの性能を検証し、対応する `/out/metrics`（性能指標）と `(log)` を生成していることを確認します。

Reviewer: ここに Lineage Graph の画像を貼ってください。

knit run show --log $evaluate_run_id
```

### 7.8.（任意）指標レポートのダウンロード
```bash
knit data pull -x $evaluate_metrics_knit_id ./out
```
このコマンドは、評価指標レポートを Knitfab からダウンロードし、`./out` ディレクトリに保存します。

## ステップ8: パイプライン構造とリネージグラフの確認
このステップでは、構築したパイプライン構造を可視化し、データ依存関係を確認します。

### 8.1. Plan グラフの生成（パイプライン構造）
Plan グラフは、パイプラインの全体像を把握し、初期学習、増分学習、検証などの異なる Plan 間の依存関係を示します。

次のコマンドを実行して、Plan グラフの PNG 画像を生成してください。`${plan_id}` は、該当する Plan Id(`$train_plan_id`、`$evaluate_plan_id`) に置き換えてください。

```bash
knit plan graph -n all ${plan_id} | dot -Tpng > plan-graph.png
```

> 以下のようにパイプライン構造を表す Plan グラフが生成されます。
> ![学習パイプライン構造](./graphs/plan_graphs/news-classification-plan-graph.svg)
> **Fig. 1:** 学習パイプライン構造

Reviewer: ```suggestion
以下のようにパイプライン構造を表す Plan グラフが生成されます。

![学習パイプライン構造](./graphs/plan_graphs/news-classification-plan-graph.svg)

**Fig. 1:** 学習パイプライン構造
```

段落を分けましょう。
さもないと、画像と前後のテキストが 1 行にまとめられてしまいます。

グラフはタスク間の依存関係を以下のように示します。

- **訓練タスクと評価タスクの関係:**
  - `train_plan` が、 `evaluate_plan` を下流タスクとして示していることを確認します。
  - これは次のことを意味します。
    - 学習タスクの後に検証が実行されること。
    - 検証タスクが学習タスクから生成された `/out/model` を受け取ること。

### 7.2. リネージグラフの生成（データフロー）

リネージグラフは、パイプライン内のデータと成果物の流れを遡ります。

以下のコマンドを実行して、リネージグラフの PNG 画像（`lineage-graph.png`）を生成してください。`${knit_id}` は、該当する Knit Id (`$train_model_knit_id`、`$evaluate_metrics_knit_id`) に置き換えてください。

```bash
knit data lineage -n all ${knit_id} | dot -Tpng > lineage-graph.png
```

> 以下のようにリネージグラフが生成されます。
> ![学習パイプラインのリネージグラフ](./graphs/lineage_graphs/news-classification-lineage-graph.svg)
> **Fig. 2:** 学習データと成果物のリネージグラフ

Reviewer: ```suggestion
以下のようにリネージグラフが生成されます。

![学習パイプラインのリネージグラフ](./graphs/lineage_graphs/news-classification-lineage-graph.svg)

**Fig. 2:** 学習データと成果物のリネージグラフ
```

段落を分けましょう。
さもないと、画像と前後のテキストが 1 行にまとめられてしまいます。
